packages/xl-ai/src/prosemirror/changeset.ts

```diff
@@ -0,0 +1,340 @@
+import { getNodeById, PartialBlock, updateBlockTr } from "@blocknote/core";
+import {
+  Change,
+  ChangeSet,
+  simplifyChanges,
+  TokenEncoder,
+} from "prosemirror-changeset";
+import { Node } from "prosemirror-model";
+import { ReplaceStep, Transform } from "prosemirror-transform";
+
+type CustomChange = Change & {
+  type?: "mark-update" | "node-type-or-attr-update";
+};
+
+/**
+ * Adds missing changes to the changes array.
+ * This is needed because prosemirror-changeset may miss some changes,
+ * such as marks or node attributes.
+ *
+ * NOTE: we might be able to replace this code with a custom encoder
+ * (this didn't exist yet when this was written)
```

Comment:
Yea, I think the whole custom encoder thing is just for this. Probably okay for now, but could be simplified